Change the autoload process

aledeg opened the pull request:
Changes proposed in this pull request:

- Change autoloading to accomodate namespaces

How to test the feature manually:

1. Use the application to check than everything is working as before.

Pull request checklist:

- [x] clear commit messages
- [x] code manually tested
- [ ] unit tests written (optional if too hard)
- [ ] documentation updated

[Additional information can be found in the documentation](https://github.com/FreshRSS/FreshRSS/tree/master/docs/en/developers/04_Pull_requests.md).

~~Before, the autoload process was too restricted. It was really dependant on our
code tree. It was hard to add more classes to be loaded automatically. On top
of that, it did not support autoloading classes following the PSR-4 recommendation.~~

~~Now, the autoload process is more open. It supports partially the PSR-4 recommendation,
there is no specific code to load Minz classes or PHPMailer classes. This is the
starting point to reorganize the codebase to introduce long waiting changes as seen
in #789. It would be a nice to later rework the tree, rename classes, and add namespace
in a fashion that follows the PSR-4. Then specific FRSS workarounds in the autoload
could be dropped.~~

aledeg committed:
Syntax fixes on autoloading process

The autoload process has been extracted to its own file.

Alkarex commented on lib/autoload.php:
```diff
@@ -0,0 +1,70 @@
+<?php
+
+class ClassLoader {
+    private $searchPath = [];
```
@Frenzie Do you know why Travis did not catch the space indenting here?
@aledeg Could you change to tabs like the other files?

Frenzie replied:
Isn't lib/* excluded or something? Please ping me in a couple of days if I forget to check.

Frenzie replied:
Not /*. Strange.
https://github.com/FreshRSS/FreshRSS/blob/0a2d9b3b54ee51a3965d79c68409ef045ad5834b/phpcs.xml#L6-L9

aledeg replied:
I've changed to tabs.

Alkarex commented on lib/autoload.php:
```diff
@@ -0,0 +1,70 @@
+        foreach ($this->searchPath as $path) {
+            foreach (glob($path . DIRECTORY_SEPARATOR . '*', GLOB_ONLYDIR) as $subPath) {
+                $file = $subPath . DIRECTORY_SEPARATOR . $freshrssClass . '.php';
+                if (file_exists($file)) {
```
On a second read, I am a bit worried about nested loops and multiple IO operations to load a file for which we are supposed to know the exact path

aledeg replied:
I removed the loops by adding manually the path during configuration.

Alkarex commented on lib/lib_rss.php:
```diff
@@ -15,6 +15,8 @@ function mb_strcut($str, $start, $length = null, $encoding = 'UTF-8') {
 	openlog('FreshRSS', LOG_CONS | LOG_ODELAY | LOG_PID, LOG_USER);
 }
 
+require_once LIB_PATH . DIRECTORY_SEPARATOR . 'autoload.php';
```
`/` works on all platforms and is shorter :-)

aledeg replied:
Except that if you read too fast like I did, you forget the `/` :)

Alkarex commented on lib/autoload.php:
```diff
@@ -0,0 +1,74 @@
+	/**
+	 * Find the file containing the class definition.
+	 */
+	public function findFile($class) {
```
I am not so found of *searching*  for the class on disk, when we have the possibility of directly knowing the class filename 😟

aledeg replied:
I agree. The initial problem was that it was too restricted and too dependent on the structure. It is still but in a different way.
Before, the rule was really strict on how to name the classes (ex: must end with _Controller), which give weird convention name and also discrepancies (ex: `FreshRSS_api_Controller` vs `FreshRSS_AlreadySubscribed_Exception`, search in Controllers for controllers but in Model for class with only one word after freshrss). That make no sense to me.
Now, there is only one rule, the class name has to match its file.

The idea with the namespace was to be able to match directly the file by using the namespace instead of relying on a weird convention on the class name to find the file to include. If you compare the process of including the class `Controller\ApiController` versus the class `FreshRSS_api_Controller`, it will be easier to convert to a path the later than the former.
With that in mind, we could remove most of the disk access to load the classes because we now right away where it is without relying on some convoluted method to extract the path from the class.

aledeg commented:
@Alkarex I am trying a different approach. But I have a blank page all the time.
And nothing is shown in the log. Do you have any ideas on how I could proceed to see more than nothing?
Thank you

Alkarex commented:
@aledeg Yes, just before `function classAutoloader($class) {`, add the following three lines:

```php
error_reporting(-1);
ini_set('display_startup_errors', 1);
ini_set('display_errors', 'On');
```

Gives something like that when a class does not exist:

```
sudo docker exec freshrss ./cli/export-opml-for-user.php --user freshrss
FreshRSS exporting OPML for user “freshrss”…

Warning: include(/var/www/FreshRSS/app/Models/EntryDAO.php): Failed to open stream: No such file or directory in /var/www/FreshRSS/lib/lib_rss.php on line 42

Warning: include(): Failed opening '/var/www/FreshRSS/app/Models/EntryDAO.php' for inclusion (include_path='.:/usr/share/php8') in /var/www/FreshRSS/lib/lib_rss.php on line 42

Fatal error: Uncaught Error: Class "FreshRSS_EntryDAO" not found in /var/www/FreshRSS/app/Models/Factory.php:37
Stack trace:
#0 /var/www/FreshRSS/app/Services/ExportService.php(32): FreshRSS_Factory::createEntryDao()
#1 /var/www/FreshRSS/cli/export-opml-for-user.php(19): FreshRSS_Export_Service->__construct()
#2 {main}
  thrown in /var/www/FreshRSS/app/Models/Factory.php on line 37
PHP Warning:  include(/var/www/FreshRSS/app/Models/EntryDAO.php): Failed to open stream: No such file or directory in /var/www/FreshRSS/lib/lib_rss.php on line 42
PHP Warning:  include(): Failed opening '/var/www/FreshRSS/app/Models/EntryDAO.php' for inclusion (include_path='.:/usr/share/php8') in /var/www/FreshRSS/lib/lib_rss.php on line 42
PHP Fatal error:  Uncaught Error: Class "FreshRSS_EntryDAO" not found in /var/www/FreshRSS/app/Models/Factory.php:37
Stack trace:
#0 /var/www/FreshRSS/app/Services/ExportService.php(32): FreshRSS_Factory::createEntryDao()
#1 /var/www/FreshRSS/cli/export-opml-for-user.php(19): FreshRSS_Export_Service->__construct()
#2 {main}
  thrown in /var/www/FreshRSS/app/Models/Factory.php on line 37
```

aledeg commented:
@Alkarex I can not spot the error.
I am trying to generate a cache file with all FreshRss classes and then load it. This will avoid to make IO calls. The cache file is generated properly but it seems I can not use it.
I do not know why.
If you could help me spot the missing element, that would be great!

aledeg commented:
I've changed the organization and I've spotted the error!

aledeg commented:
@Alkarex I think my work is done here. I kept the filesystem search for freshrss classes but I cached it. Now, it is search once and cached. Later, the cached version is retrieved automatically.
Let me know what you think.

aledeg commented:
@Alkarex I was working on that long overdue PR and I cannot understand why phpstan tests are failing. By branch is up-to-date with the parent branch and the failing tests are somehow unrelated to my changes.
Do you have any ideas?

Alkarex commented on lib/Autoload/UniversalClassLoader.php:
```diff
@@ -0,0 +1,56 @@
+	 * Find the file containing the class definition.
+	 */
+	protected function findFile($class) {
+		foreach ($this->registeredPath as $path) {
```
I am still not a fan of searching on disk, something very slow (especially since it needs to be done for every PHP file we load), when we already know where it is...

aledeg replied:
I can add caching on that as well.

Alkarex replied:
P.S. I know you have tried to explain higher up the rationale, but I still don't get it. It is significantly longer code (harder to understand, at least to me) with loops and disk scans, and it does not look like it solves any current problem, does it?

Alkarex commented:
> @Alkarex I was working on that long overdue PR and I cannot understand why phpstan tests are failing. By branch is up-to-date with the parent branch and the failing tests are somehow unrelated to my changes. Do you have any ideas?

@aledeg I guess it is because PHPStan fails to use the new autoloading, in particular loading `./data/cache/18dd0b692afe4afac65809e7eb081f9f0269e841.class.php'` leading to many errors

aledeg commented:
@Alkarex I've refactor the existing code and removed what I have done previously.
I extracted the autoload in its own file so we could include it easily in composer or phpunit configuration if needed.

aledeg commented:
@Alkarex I've added the use of the autoloading in the I18n section of the cli tool. We do not need to play around with `include` to make the thing working.
I've also added a namespace for said section.
